api/lib/domain/models/Challenge.js

```diff
@@ -87,6 +87,23 @@ class Challenge {
     this.competenceId = competenceId;
   }
 
+  _orderAttachments(attachments) {
+    if (!attachments) {
+      return [];
+    }
+    const microsoftFormats = ['docx', 'xlsx', 'pptx'];
+    const orderedAttachments = [];
+    attachments.forEach((attachment) => {
+      const format = attachment.split('.').pop();
+      if (microsoftFormats.includes(format)) {
+        orderedAttachments.unshift(attachment);
+      } else {
+        orderedAttachments.push(attachment);
+      }
+    });
+    return orderedAttachments;
```

Comment:
J'aurait utilisé la fonction `sort` pour cela:
```suggestion
      return attachments.sort((a, b) => {
      const formatB = b.split('.').pop();
      if (microsoftFormats.includes(formatB)) {
        return 1;
      }
      return -1;
    });
```

Reply:
@francois2metz @celineung Nooooon ! 😱 

La méthode `sort()` [mute le tableau sur lequel elle est appelée](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) or rien n'autorise le constructeur de `Challenge` à modifier ce tableau qui lui a été passé ici 
https://github.com/1024pix/pix/blob/282e18ca6b0f522a0f7c079b379fcf2798dfbe72/api/lib/infrastructure/repositories/challenge-repository.js#L95

et qui est une référence directe à un objet du cache mémoire du référentiel. Autrement dit, à chaque appel au `challenge-repository` on modifie maintenant un objet qui est persistant en mémoire.

De plus, le comparateur passé ici à `sort` ne garantit pas un ordre répétable : en effet il indique par exemple que `ods` doit venir après `xls` mais aussi que `ods` vient après `odp` et… que `odp` vient après `ods` 😕 

Dernière remarque, ce code est appelé très fréquemment, n'est-ce pas dommage de refaire systématiquement un travail de tri qui pourrait être fait : soit au niveau de LCMS si on considère que l'ordre des pièces jointes est significatif pour la définition de l'épreuve ; soit au niveau de `mon-pix` s'il ne s'agit que d'une question de présentation ?